DM-46599: Deprecate old query system behavior

dhirving opened the pull request:
Deprecate behavior from the old query system that will no longer be supported when we switch to the new implementation.

## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

dhirving committed:
Deprecate regular expression searches

dhirving committed:
Deprecate materialize

dhirving committed:
Deprecate datasetType parameter to queryCollections

dhirving committed:
Deprecate query governor validation

dhirving committed:
Deprecate no-check option to query-dimension-records

dhirving committed:
Deprecate offset parameter to query limit

dhirving committed:
Deprecate empty dataset type/collection in query-datasets

dhirving committed:
Deprecate butler.collections as list of defaults

dhirving committed:
Fix tests for deprecation of butler.collections

dhirving committed:
Fix spurious warnings about regex

This was also flagging globs when it shouldn't have been

dhirving committed:
Remove unnecessary queryCollections datasetType

dhirving committed:
Deprecate htm/healpix data IDs

In the new query system, this functionality can be expressed more clearly using region overlaps.

dhirving committed:
Suppress deprecation warnings in registry tests

dhirving committed:
Suppress misc deprecation warnings in tests

dhirving committed:
Handle deprecation in query-datasets test

Collections and dataset type arguments are becoming mandatory, so add them to tests that were using the default 'all' values.

dhirving committed:
Fix tests for missing RemoteButler warning

dhirving committed:
Tweak deprecation messages

dhirving committed:
Add towncrier

timj commented on doc/changes/DM-46599.removal.md:
```diff
@@ -0,0 +1,19 @@
+
+The `offset` argument to `limit()` for `Registry.queryDataIds` and `Registry.queryDimensionRecords` result objects is now deprecated.
+
+The `--offset` option for `butler query-data-ids` and `butler-query-datasets` is now deprecated.
```
It looks like `butler query-data-ids` raises when `--offset` is used and `query-dimension-records` ignores it. Maybe we add a raise to the latter as well? Or remove the raise from the former and add deprecation warnings instead noting that the option is a no-op? I think this happened during the migration to the new query system.

dhirving replied:
I didn't realize that these had been switched over to the new query system... theoretically shouldn't have happened until after this deprecation.

Probably better to raise than silently ignore the option, so I'll add an exception to `query-dimension-records`.

timj reviewed (approved):
Looks great.

dhirving committed:
Add exception for query-dimension-records offset

The --offset parameter to query-dimension-records was changed to be silently ignored in DM-45556.  It can't be brought back without going back to the old query system.

We are removing this entirely soon, and it's better to flag the problem to the user than silently ignore what they asked for.

Also added a little more information to a similar information that was being thrown from query-data-ids.
